pytools/persistent_dict.py

```diff
@@ -549,112 +447,122 @@ def __init__(self, identifier: str,
             if sys.platform == "darwin" and os.getenv("XDG_CACHE_HOME") is not None:
                 # platformdirs does not handle XDG_CACHE_HOME on macOS
                 # https://github.com/platformdirs/platformdirs/issues/269
-                cache_dir = join(os.getenv("XDG_CACHE_HOME"), "pytools")
+                container_dir = join(os.getenv("XDG_CACHE_HOME"), "pytools")
             else:
-                cache_dir = platformdirs.user_cache_dir("pytools", "pytools")
+                container_dir = platformdirs.user_cache_dir("pytools", "pytools")
 
-            container_dir = join(
-                    cache_dir,
-                    "pdict-v4-{}-py{}".format(
-                        identifier,
-                        ".".join(str(i) for i in sys.version_info)))
+        self.filename = join(container_dir, f"pdict-v5-{identifier}"
+                             + ".".join(str(i) for i in sys.version_info)
+                             + ".sqlite")
 
         self.container_dir = container_dir
-
         self._make_container_dir()
 
-    @staticmethod
-    def _warn(msg: str, category: Any = UserWarning, stacklevel: int = 0) -> None:
-        from warnings import warn
-        warn(msg, category, stacklevel=1 + stacklevel)
-
-    def store_if_not_present(self, key: K, value: V,
-                             _stacklevel: int = 0) -> None:
-        """Store (*key*, *value*) if *key* is not already present."""
-        self.store(key, value, _skip_if_present=True, _stacklevel=1 + _stacklevel)
-
-    def store(self, key: K, value: V, _skip_if_present: bool = False,
-              _stacklevel: int = 0) -> None:
-        """Store (*key*, *value*) in the dictionary."""
-        raise NotImplementedError()
-
-    def fetch(self, key: K, _stacklevel: int = 0) -> V:
-        """Return the value associated with *key* in the dictionary."""
-        raise NotImplementedError()
-
-    @staticmethod
-    def _read(path: str) -> V:
-        from pickle import load
-        with open(path, "rb") as inf:
-            return load(inf)
-
-    @staticmethod
-    def _write(path: str, value: V) -> None:
-        from pickle import HIGHEST_PROTOCOL, dump
-        with open(path, "wb") as outf:
-            dump(value, outf, protocol=HIGHEST_PROTOCOL)
-
-    def _item_dir(self, hexdigest_key: str) -> str:
-        from os.path import join
+        # isolation_level=None: enable autocommit mode
+        # https://www.sqlite.org/lang_transaction.html#implicit_versus_explicit_transactions
+        self.conn = sqlite3.connect(self.filename, isolation_level=None)
 
-        # Some file systems limit the number of directories in a directory.
-        # For ext4, that limit appears to be 64K for example.
-        # This doesn't solve that problem, but it makes it much less likely
+        self.conn.execute(
+            "CREATE TABLE IF NOT EXISTS dict "
+            "(keyhash TEXT NOT NULL PRIMARY KEY, key_value TEXT NOT NULL)"
+            )
```

Comment:
Explain somewhere that it's a deliberate choice to not store all values with a certain key, based on the assumption that key conflicts are supremely unlikely, and if they occur, almost always due to a bug.

Reply:
What do you think of b4b623a?